Add source editor component

edmundoa opened the pull request:
This PR moves the source code editor we used in the pipelines to the web interface, so other plugins can reuse it if they need it.

I manage to make a few improvements in there, hopefully they will improve the experience of using it:
- Updated to latest version of ReactAce
- Fixed issue with undo command after load: https://github.com/Graylog2/graylog-plugin-pipeline-processor/issues/224. I'm not fully satisfied with the fix, so I'll try to do a better fix and get it upstream
- Allow editor to be user-resizable, fixing https://github.com/Graylog2/graylog-plugin-pipeline-processor/issues/229
- Add read-only flag, so we can also pre-visualize code with the component
- Add toolbar to make more visibles some actions that can be done in the editor. For now only copying, pasting*, undo, and redo are available.

Bonus: Add dark theme support! 
<img width="993" alt="screen shot 2017-12-13 at 17 36 42" src="https://user-images.githubusercontent.com/716185/33950293-5707a754-e02c-11e7-88b2-398410560938.png">

*Dispatching a pasting event from JS doesn't really work for privacy reasons, so the user must trigger an action in the OS. The button therefore only shows a small tooltip indicating it can be done in the Edit menu.

edmundoa commented:
PS: For now the dark theme is only supported when specified via props. I guess we can do something better with that once we do some other improvements 🙂

kroepke commented on graylog2-web-interface/src/components/common/SourceCodeEditor.jsx:
```diff
@@ -0,0 +1,205 @@
+     * `[{ row: 0, column: 2, type: 'error', text: 'Some error.'}]`
+     * The type value must be one of `error`, `warning`, or `info`.
+     */
+    annotations: PropTypes.array,
```
Is it worth using a `PropTypes.shape` here or do we not bother because this typically comes from the backend anyway?

edmundoa replied:
I'm most of the time against those shape objects, as you need to define the whole schema of an object (or an array of objects in this case) in a way that is not very readable and it's easy to make you waste more time removing warnings than helping catching errors. Additionally, in this case the shape is highly dependant on ace, and they don't even have good documentation for it. For instance, the `column` property didn't seem to have any effect when I tested it, as all annotations appear in the gutter on the right of the editor. I also had to go to the source code to see what types were supported.
In my opinion, having an example and the vales the `type` property accepts is good enough.

kroepke replied:
Works for me :)

kroepke commented on graylog2-web-interface/src/components/common/SourceCodeEditor.jsx:
```diff
@@ -0,0 +1,205 @@
+                       editorProps={{ $blockScrolling: 'Infinity' }}
+                       focus={this.props.focus}
+                       fontSize={this.props.fontSize}
+                       mode="text"
```
Do we want to extract this into an optional prop, too?
Might come in handy if we ever use this to display/edit JSON or other modes.

edmundoa replied:
The main problem is that each mode requires an `import` from `brace`. I couldn't find a way of importing all the modes brace provides, and I'm unsure if it's even safe to do so. I can play a bit more with it and see how it goes. Otherwise I would vote for extending the component when we need it.

kroepke replied:
Yeah makes sense, let's not overcomplicated it for now.

kroepke commented on graylog2-web-interface/src/components/common/SourceCodeEditor.md:
````diff
@@ -0,0 +1,67 @@
+                  value={code} />
+```
+
+Non-resizable editor without toolbar and with custom height and width:
````
States that the editor isn't resizable, when in fact it is.

edmundoa replied:
That was me changing props and texts last minute 😅 Will fix it.

edmundoa commented:
@kroepke addressed your comments and other things I saw:
- Now we support JSON, Lua, Markdown, and YAML languages with the `mode` prop
- Improved the logic to disabled copy and paste buttons
- Avoided resizing editor and text selection when they don't make sense
- Added a note regarding uncontrolled usages of the component, also updated examples to always keep the state themselves. React-Ace is supposed to keep the text value if you don't update it by hand, but that seems to fail in some situations

edmundoa commented:
I rebased this against master and also removed the hack to reset the undo history. The issue was fixed upstream.

edmundoa commented:
The failing test is unrelated to the changes, but to the new year :tada: I'll submit another PR fixing it.

bernd commented on graylog2-web-interface/src/components/common/SourceCodeEditor.jsx:
```diff
@@ -0,0 +1,205 @@
+                                 onSuccess={this.focusEditor}
+                                 text={this.state.selectedText}
+                                 buttonTitle="Copy (Ctrl+C / &#8984;C)"
+                                 disabled={this.isCopyDisabled} />
```
This passes the `isCopyDisabled` function instead of executing it.

edmundoa replied:
Good finding! This got broken in one of the changes I did.

bernd commented on graylog2-web-interface/src/components/common/SourceCodeEditor.jsx:
```diff
@@ -0,0 +1,205 @@
+                                 buttonTitle="Copy (Ctrl+C / &#8984;C)"
+                                 disabled={this.isCopyDisabled} />
+                <OverlayTrigger placement="top" trigger="click" overlay={overlay} rootClose>
+                  <Button bsStyle="link" bsSize="sm" title="Paste (Ctrl+V / &#8984;V)" disabled={this.isPasteDisabled}>
```
This passes the `isPasteDisabled` function instead of executing it.
